### DISCUSSION TITLE
Excess solar / minimising losses with early morning force discharge?

### DISCUSSION THREAD

**Discussion by matknowles** [springfall2008/batpred#1076] [Ideas]
Hello, I've had an idea to perhaps reduce losses from excess solar going to earth.

Say I have a 5kW inverter but am gathering 6kW solar; it's outputting 5kW max AC to house/grid then dumping the excess if my batteries are fully charged.
If my understanding is correct, the inverter can, in addition to that 5kW AC output, charge my batteries at up to 3.6kW DC (assuming 8.6kW solar generation)

As I'm on Intelligent Octopus Go, I'm taking advantage of fully charging the batteries 23.30 - 05.30 which also takes advantage of import prices at 7.5p and export prices at 15p. The issue arises in that the batteries discharge only slightly from powering the house from 05.30 to sunrise; the strong solar for the coming day then quickly fills the batteries, then as they hit 100% early morning, the output is then capped at 5kW AC as the batteries are full.

Therefore if Predbat knows it's going to be such a sunny day that the batteries charge and stay at 100%, perhaps a routine could be implemented whereby the batteries force discharge after 05.30 for as long as possible whilst still aiming to either keep enough battery for the house for the day, or to just get to ~100% with the available solar. (I'm waffling there but hopefully my point comes across)

Apologies if this isn't a new idea or not possible, it just occurred to me the other day seeing our solar "go to waste" as it were (we were generating ~7.8kW with full batteries; exporting at ~3kW AC and using ~2kW AC in the house, "wasting" ~2.8kW)

Thanks a lot

**Comment by cyberkryten** [comment]
I have very similar issues.

I have a 3.6kW G100 Export limit, which I've set and the house uses 1-2kW but generation predictions (and actual) are 6.5kW (and this will increase in the next months).  I have 3x9.7kWh batteries and the plan shows these being fully charged in the IOG 7.5p slot overnight.

I was expecting the system to not fully charge the batteries on these sunny days, as there's clearly excess solar in the plan that cannot be used or exported.  

If I didn't have the export limit, then the correct thing would be to do what it is doing and charge the batteries fully @ 7.5p and export all generated solar @ 15p, but the export limit changes that equation because anything over approx 5kW of generation is going to waste, so I'm paying to fill the batteries when I could be adding this excess power to them for free.

It looks like the method of dealing with this is to force discharge through the day, so aiming to take the battery from 100% down to 95% to allow some headroom for the excess solar, however when there is an export limit that's not possible because there's already sufficient solar to max that out, meaning the battery state never reduces.

I've played with several settings on SoC but can't figure out how to leave a margin and assumed I'd missed something or got the config wrong, but I'm guessing that the system just doesn't account for the excess solar/export limits?

**Comment by matknowles** [comment]
Yeah I really can't think of any other way of handling it other than scheduling a force discharge between the end of cheap rate and when the bulk of the solar kicks in, or perhaps juggling the cheap charge with intermittent discharges (like the set_discharge_during_charge?)

I guess the issue is ensuring that the solcast prediction is sure enough that there'll be sufficient solar to cause this problem in the first place, so maybe when even the pessimistic solar forecast also shows a high level for the day.

**Comment by cyberkryten** [comment]
I've only been running this for a few days, but I wonder if Issue #1101 is related?  

It suggests that something broke with the SoC calcs in a recent version, resulting in a high (max?) SoC overnight.  That would explain what I'm seeing and probably what you are too - fully charging overnight without accounting for the following day's solar excess over the inverter capacity/export limit.

If you've been running for a while (from 7.18.3) then maybe you can see if your plan has changed for sunny days since then to help track it down?

**Comment by matknowles** [comment]
I've been using predbat for a few months now and it's been generally great. Recently there've been a few nasty bugs, but they're being squashed within a few days it seems. This topic isn't realy an issue or problem it's just a suggestion to further improve gains.

**Reply by cyberkryten** [reply]
Ah, sorry, I think I misunderstood the request, just assumed it was the same as the other issue reported

**Comment by DrJohnM61** [comment]
> Hello, I've had an idea to perhaps reduce losses from excess solar going to earth.
> 
> Say I have a 5kW inverter but am gathering 6kW solar; it's outputting 5kW max AC to house/grid then dumping the excess if my batteries are fully charged. If my understanding is correct, the inverter can, in addition to that 5kW AC output, charge my batteries at up to 3.6kW DC (assuming 8.6kW solar generation)
> 
> As I'm on Intelligent Octopus Go, I'm taking advantage of fully charging the batteries 23.30 - 05.30 which also takes advantage of import prices at 7.5p and export prices at 15p. The issue arises in that the batteries discharge only slightly from powering the house from 05.30 to sunrise; the strong solar for the coming day then quickly fills the batteries, then as they hit 100% early morning, the output is then capped at 5kW AC as the batteries are full.
> 
> Therefore if Predbat knows it's going to be such a sunny day that the batteries charge and stay at 100%, perhaps a routine could be implemented whereby the batteries force discharge after 05.30 for as long as possible whilst still aiming to either keep enough battery for the house for the day, or to just get to ~100% with the available solar. (I'm waffling there but hopefully my point comes across)
> 
> Apologies if this isn't a new idea or not possible, it just occurred to me the other day seeing our solar "go to waste" as it were (we were generating ~7.8kW with full batteries; exporting at ~3kW AC and using ~2kW AC in the house, "wasting" ~2.8kW)
> 
> Thanks a lot

I have the same issue. My DNO restricts my exports to 5kW. Even discharging the battery fully in the morning will result in a full battery (13.5kWh AIO) early afternoon and then waisted solar. Not only would I like to see an option to discharge to the grid in the morning (based on solar predication), but also being able to throttle the battery charge to a number (eg, 2.5kW) so that the max is going to the grid but the battery does not top out and force clipping. (eg, 500W to the house, 2.5kW to the battery and 5Kw to the grid = 8Kw [my inverters max]). The battery will still get full after 5 or so hours and I will export the max possible otherwise, the battery gets 6kWh for two hours and then the excess solar clips from 7.5kWh available to the grid to just 5kWh.

**Comment by matknowles** [comment]
As a proof of concept, I force exported for an additional half hour or so this morning as I saw the solar prediction was great for the day. It gave me some headroom in the battery (to about 85%) which was then utilised when the solar ramped up for the day. Attached is a screenshot proving my original point as well, that solar over 5kW can be additionally used on the DC side to charge the battery.
![Screenshot_20240527-114435](https://github.com/springfall2008/batpred/assets/153369667/346b0047-e804-4e91-bbf4-df8188500fbe)


The main issue is the unpredictability of the solar forecast. I think if Predbat can see the probabilities for pessimistic/average and optimistic (or however the 3 averages work) are all showing lots of solar, then this routine can be implemented.

So cheap charging to 100% until 0530.
Force discharging for perhaps 2-3 hours (Accounting for house load/upcoming solar as usual)
Solar >5kW can be more effectively used.
This also has the benefit of helping the people as above who have export limitations in place.
